File: src/events/session.ts
"use client";

import { useEffect } from "react";
import { useStore } from "react-redux";

import type { AppSession } from "@/cmd/session";
import { SESSION_CHANGED_EVENT } from "@/config/window-events";
import type { AppStore } from "@/store";
import { changeCurrentLanguageAction } from "@/store/modules/app";
import { tauriOn } from "@/utils/tauri-event";

export function applySessionToStore(store: AppStore, session: AppSession) {
  const lang = session.currentLanguage;
  if (lang === "cn" || lang === "en") {
    store.dispatch(changeCurrentLanguageAction(lang));
  }
}

/** Rust `session/changed` → 各 Webview 本地 Redux */
export function useSessionChangedSubscription() {
  const store = useStore() as AppStore;

  useEffect(() => {
    return tauriOn<AppSession>(SESSION_CHANGED_EVENT, (event) => {
      applySessionToStore(store, event.payload);
    });
  }, [store]);
}

File: src-tauri/src/cmd/session.rs
use tauri::AppHandle;

use crate::context::session::{self, AppSession};

#[tauri::command]
pub fn get_app_session(app: AppHandle) -> Result<AppSession, String> {
    let snapshot = session::get_session(&app)?;
    crate::log_info!(
        "cmd.session.get_app_session ok lang={}",
        snapshot.current_language
    );
    Ok(snapshot)
}
